File: trees/sum_root_the_leaf_numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int result = 0;
        stack<pair<TreeNode*, int>> s;
        s.push({root,0});

        while (!s.empty()) {
            pair<TreeNode*, int> p = s.top();
            s.pop();

            TreeNode* node = p.first;
            if (node) {
                int current = p.second * 10 + node->val;
                if (!node->left && !node->right) result += current;
                else {
                    s.push({node->left, current});
                    s.push({node->right, current});
                }
            }
        }

        return result;
    }
};

File: trees/sum_root_the_leaf_numbers.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int SumNumbers(TreeNode root) {
        int result = 0;

        Stack<(TreeNode, int)> s = [];
        s.Push((root, 0));

        while (s.TryPop(out var item)) {
            var (node, val) = item;
            if (node != null) {
                int current = val * 10 + node.val;
                if (node.left == null && node.right == null) result += current;
                else {
                    s.Push((node.left, current));
                    s.Push((node.right, current));
                }
            }
        }

        return result;
    }
}
